### PR TITLE
ndslice: allocation utilities

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

**Comment by wilzbach** [comment]
Typo in the pr name ;-)

**Comment by 9il** [comment]
@John-Colvin @wilzbach please review allocators for ndslice

File: source/mir/ndslice/slice.d [hunk 1]
```diff
@@ -583,6 +491,195 @@ pure nothrow unittest
 }
 
 /++
+Creates an array and an n-dimensional slice over it.
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d:494]: ~~nitpick: 'a' n-dim.~~
Update: sorry. My mistake

File: source/mir/ndslice/slice.d [hunk 2]
```diff
@@ -583,6 +491,195 @@ pure nothrow unittest
...
+Params:
+    lengths = list of lengths for each dimension
+Returns:
+    n-dimensional slice
++/
+Slice!(Lengths.length, Select!(replaceArrayWithPointer, T*, T[]))
+createSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d:502]: just out of interest? why no double quotes? `"`
Maybe we should start a style guide then ...

Comment (9il) [inline source/mir/ndslice/slice.d:502]: Unification with other crazy code

File: source/mir/ndslice/slice.d [hunk 3]
```diff
@@ -583,6 +491,195 @@ pure nothrow unittest
...
+Params:
+    alloc = allocator
+    lengths = list of lengths for each dimension
+    init = default value for array initialization
+Returns:
+    a structure with fields `array` and `slice`
++/
+auto makeSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
+    Allocator,
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d:544]: something that I was asking myself when working on the combinatorics module - is there any  template check for the Allocator api? If not - we might want to create one.

Comment (9il) [inline source/mir/ndslice/slice.d:544]: No, please, just less imports

File: source/mir/ndslice/slice.d [hunk 4]
```diff
@@ -583,6 +491,195 @@ pure nothrow unittest
...
+}
+
+/// ditto
+auto makeSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
+    Allocator,
+    size_t N)(auto ref Allocator alloc, auto ref in size_t[N] lengths)
+{
+    import std.experimental.allocator: makeArray;
+    static struct Result { T[] array; Slice!(N, Select!(replaceArrayWithPointer, T*, T[])) slice; }
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d:558]: why don't you put this out of the scope?
It is used by the method below too.

Comment (9il) [inline source/mir/ndslice/slice.d:558]: This is template anyway, and small

**Comment by wilzbach** [comment]
> @John-Colvin @wilzbach please review allocators for ndslice

Wow this is great stuff! I tried to review as best as I can, but I still have to learn.

Do you plan to allow multi-dimensional arrays as "input"? - could be a convenient feature, e.g.

```
(new int [2][2]).sliced;
```

**Comment by 9il** [comment]
> Do you plan to allow multi-dimensional arrays as "input"? - could be a convenient feature, e.g.

This is impossible because ndslice requires single range.
In the same time you can do assignment `aSlice = anArray`. So first you need to create a slice, and then assign an array to this slice. Looks like I need to add function, witch would return a shape for common D array

**Pull request merged**